Enable fused optimizer for DP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98270

Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

rohan-varma committed:
Enable fused optimizer for DP



Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1464,9 +1464,9 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
     if (!gradient_as_bucket_view_) {
-      if (set_grads_to_none_) {
+      if (optim_in_backward_) {
+        // Return early has optimizer has already run.
```
nit: Maybe typo?
```suggestion
        // Return early since optimizer has already run.
```

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1486,8 +1486,8 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
-        if (set_grads_to_none_) {
-          grad.reset();
+        if (optim_in_backward_) {
+          // Return early has optimizer has already run.
```
Same as above: maybe typo?

awgu reviewed (approved):
LGTM. There is one real lint error.

rohan-varma committed:
Update on "Enable fused optimizer for DP"


Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable fused optimizer for DP"


Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable fused optimizer for DP"


Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable fused optimizer for DP"


Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable fused optimizer for DP"


Differential Revision: [D42714482](https://our.internmc.facebook.com/intern/diff/D42714482/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D42714482/)!

[ghstack-poisoned]
